scripts/nix-build.in

```diff
@@ -191,17 +191,34 @@ for (my $n = 0; $n < scalar @ARGV; $n++) {
         $n++;
         die "$0: ‘$arg’ requires an argument\n" unless $n < scalar @ARGV;
         my $interpreter = $ARGV[$n];
+        my $execArgs = "";
+
         # Überhack to support Perl. Perl examines the shebang and
         # executes it unless it contains the string "perl" or "indir",
         # or (undocumented) argv[0] does not contain "perl". Exploit
         # the latter by doing "exec -a".
-        my $execArgs = $interpreter =~ /perl/ ? "-a PERL" : "";
+
+        if ($interpreter =~ /perl/) {
+            $execArgs = "-a PERL";
+        }
+
         sub shellEscape {
             my $s = $_;
             $s =~ s/'/'\\''/g;
             return "'" . $s . "'";
         }
-        $envCommand = "exec $execArgs $interpreter $script ${\(join ' ', (map shellEscape, @savedArgs))}";
+
+        if ($interpreter =~ /ruby/) {
+
+            # Hack for Ruby. Ruby also examines the shebang. It tries to
+            # read the shebang to understand which packages to read from. Since
+            # this is handled via nix-shell -p, we wrap our ruby script execution
+            # in an eval.
+            my $cmd = "$interpreter -e 'eval(File.read(\"$script\"), nil, \"$script\")'";
+            $envCommand = "exec $execArgs $cmd ${\(join ' ', (map shellEscape, @savedArgs))}";
```

Comment:
I'll use your example but the third argument to ruby eval should set $0 correctly. Maybe. You know i'm not exactly sure. Load is certainly cleaner, thank you.